File: ai_pod_cli/source_codec.py
"""Strict ActUnit-compatible create/path/content subset; decoding never writes files."""

import re
from ai_pod_cli.instruction_protocol import InstructionSyntaxError


def _valid_characters(value: str) -> None:
    for char in value:
        code = ord(char)
        if not (code in (9, 10, 13) or 0x20 <= code <= 0xD7FF
                or 0xE000 <= code <= 0xFFFD or 0x10000 <= code <= 0x10FFFF):
            raise ValueError("Invalid XML character in source artifact")


def _entities(value: str) -> str:
    if "]]>" in value:
        raise ValueError("CDATA terminator requires split CDATA sections")
    named = {"amp": "&", "lt": "<", "gt": ">", "quot": '"', "apos": "'"}

    def replace(match):
        entity = match.group(1)
        if entity in named:
            return named[entity]
        if entity and re.fullmatch(r"(?:#[0-9]+|#x[0-9a-fA-F]+)", entity):
            try:
                code = int(entity[2:], 16) if entity.startswith("#x") else int(entity[1:])
                result = chr(code)
                _valid_characters(result)
                return result
            except (ValueError, OverflowError):
                pass
        raise ValueError("Invalid XML entity")

    return re.sub(r"&([^;]*);|&", replace, value)


def encode_source_artifact(path: str, content: str) -> str:
    _valid_characters(path)
    _valid_characters(content)
    path = path.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
    content = content.replace("]]>", "]]]]><![CDATA[>")
    return f"<create><path>{path}</path><content><![CDATA[{content}]]></content></create>"


def decode_source_artifact(text: str, expected_path: str | None = None) -> dict[str, str]:
    if not isinstance(text, str):
        raise ValueError("Source generation must return XML text")
    if len(text) > 2_000_000:
        raise ValueError("Source artifact response exceeds 2,000,000 characters")
    _valid_characters(text)
    position = 0

    def fail(reason, offset=None, code="invalid_instruction"):
        raise InstructionSyntaxError(reason, position if offset is None else offset, code)

    def whitespace():
        nonlocal position
        while position < len(text) and text[position].isspace():
            position += 1

    def tag():
        nonlocal position
        match = re.match(r"<(\/?)(?:｜DSML｜)?([A-Za-z_][A-Za-z0-9_.-]*)\s*>", text[position:])
        if match is None:
            fail("Expected plain XML tag; attributes and declarations are not allowed")
        position += match.end()
        return match[1] + match[2]

    def scalar(name):
        nonlocal position
        chunks = []
        while position < len(text):
            marker = next((item for item in ("<![CDATA[", "<｜DSML｜CDATA[")
                           if text.startswith(item, position)), None)
            if marker:
                start = position + len(marker)
                end = text.find("]]>", start)
                if end < 0:
                    fail(f"Incomplete CDATA section: close with ]]> (not ]]]), then </{name}>", code="missing_cdata_end")
                chunks.append(text[start:end])
                position = end + 3
            elif text[position] == "<":
                start = position
                found = tag()
                if found != f"/{name}":
                    fail(f"Expected </{name}>, found <{found}>. Nested XML operands are not allowed; use CDATA for source.", start, "mismatched_tag")
                return "".join(chunks)
            else:
                end = text.find("<", position)
                if end < 0:
                    fail(f"Incomplete XML operand <{name}>: missing </{name}>", code="unclosed_tag")
                chunks.append(_entities(text[position:end]))
                position = end
        fail(f"Incomplete XML operand <{name}>: missing </{name}>", code="unclosed_tag")

    whitespace()
    if tag() != "create":
        fail("Expected exactly one create artifact action", 0)
    fields = {}
    while True:
        whitespace()
        start = position
        name = tag()
        if name == "/create":
            break
        if name not in ("path", "content") or name in fields:
            fail(f"Unknown or duplicate artifact operand <{name}>; supply <path> and <content> exactly once", start)
        fields[name] = scalar(name)
    whitespace()
    if position != len(text):
        fail("Expected exactly one artifact; trailing actions or text are not allowed", code="multiple_instructions")
    if set(fields) != {"path", "content"}:
        fail("Artifact requires path and content")
    if expected_path is not None and fields["path"] != expected_path:
        raise ValueError(f"Artifact path must equal planned path '{expected_path}'")
    return fields

File: aipod-node/src/agent/source-codec.ts
import { INSTRUCTION_SET_PROMPT, InstructionSyntaxError } from "./instruction-protocol.js";

/** Strict ActUnit-compatible create/path/content subset, not a general XML parser.
 * Exactly one candidate artifact is decoded; no action is executed here.
 */
export interface SourceArtifact {
  path: string;
  content: string;
}

export function validCharacters(value: string): void {
  for (const character of value) {
    const code = character.codePointAt(0)!;
    if (!(code === 9 || code === 10 || code === 13 ||
      (code >= 0x20 && code <= 0xd7ff) || (code >= 0xe000 && code <= 0xfffd) || code >= 0x10000)) {
      throw new Error("Invalid XML character in source artifact");
    }
  }
}

export function decodeEntities(value: string): string {
  if (value.includes("]]>")) throw new Error("CDATA terminator must be inside split CDATA sections");
  return value.replace(/&([^;]*);|&/g, (match, entity: string | undefined) => {
    const named: Record<string, string> = { amp: "&", lt: "<", gt: ">", quot: '"', apos: "'" };
    if (entity && Object.hasOwn(named, entity)) return named[entity]!;
    if (entity && /^(?:#[0-9]+|#x[0-9a-fA-F]+)$/.test(entity)) {
      const code = entity.startsWith("#x") ? Number.parseInt(entity.slice(2), 16) : Number(entity.slice(1));
      if (code >= 0 && code <= 0x10ffff) {
        const decoded = String.fromCodePoint(code);
        validCharacters(decoded);
        return decoded;
      }
    }
    throw new Error(`Invalid XML entity '${match}'`);
  });
}

export function encodeSourceArtifact(artifact: SourceArtifact): string {
  validCharacters(artifact.path);
  validCharacters(artifact.content);
  const path = artifact.path.replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;");
  const content = artifact.content.replaceAll("]]>", "]]]]><![CDATA[>");
  return `<create><path>${path}</path><content><![CDATA[${content}]]></content></create>`;
}

export function decodeSourceArtifact(text: string, expectedPath?: string): SourceArtifact {
  if (text.length > 2_000_000) throw new Error("Source artifact response exceeds 2,000,000 characters");
  validCharacters(text);
  let position = 0;
  const fail = (reason: string, code = "invalid_instruction", at = position): never => { throw new InstructionSyntaxError(reason, at, code); };
  const whitespace = () => { while (/\s/.test(text[position] ?? "") && position < text.length) position += 1; };
  const tag = (): string => {
    const match = /^<(\/?)(?:｜DSML｜)?([A-Za-z_][A-Za-z0-9_.-]*)\s*>/.exec(text.slice(position));
    if (!match) return fail("Expected a plain XML tag; attributes, declarations and nested content are not allowed");
    position += match[0].length;
    return `${match[1]}${match[2]}`;
  };
  const scalar = (name: string): string => {
    let value = "";
    while (position < text.length) {
      const marker = text.startsWith("<![CDATA[", position) ? "<![CDATA["
        : text.startsWith("<｜DSML｜CDATA[", position) ? "<｜DSML｜CDATA[" : undefined;
      if (marker) {
        const start = position + marker.length;
        const end = text.indexOf("]]>", start);
        if (end < 0) return fail(`Incomplete CDATA section: close with ]]> (not ]]]), then </${name}>`, "missing_cdata_end");
        value += text.slice(start, end);
        position = end + 3;
      } else if (text[position] === "<") {
        const start = position, found = tag();
        if (found !== `/${name}`) return fail(`Expected </${name}>, found <${found}>. Nested XML is not allowed inside '${name}'; use CDATA for source`, "mismatched_tag", start);
        return value;
      } else {
        const end = text.indexOf("<", position);
        if (end < 0) return fail(`Incomplete XML operand <${name}>: missing </${name}>`, "unclosed_tag");
        value += decodeEntities(text.slice(position, end));
        position = end;
      }
    }
    return fail(`Incomplete XML operand <${name}>: missing </${name}>`, "unclosed_tag");
  };
  whitespace();
  if (tag() !== "create") fail("Expected exactly one create artifact action", "invalid_instruction", 0);
  const fields = new Map<string, string>();
  while (true) {
    whitespace();
    const start = position;
    const name = tag();
    if (name === "/create") break;
    if (name !== "path" && name !== "content") fail(`Unknown artifact operand '${name}'; supply path and content exactly once`, "invalid_instruction", start);
    if (fields.has(name)) fail(`Duplicate artifact operand '${name}'`, "invalid_instruction", start);
    fields.set(name, scalar(name));
  }
  whitespace();
  if (position !== text.length) fail("Expected exactly one artifact; trailing actions or text are not allowed", "multiple_instructions");
  if (!fields.has("path") || !fields.has("content")) fail("Artifact requires path and content");
  const path = fields.get("path")!;
  if (expectedPath !== undefined && path !== expectedPath) throw new Error(`Artifact path must equal planned path '${expectedPath}'`);
  return { path, content: fields.get("content")! };
}

export function sourceArtifactInstruction(path: string): string {
  return INSTRUCTION_SET_PROMPT + `Return exactly one XML-like <create> instruction with <path> and <content> operands. The path must be exactly ${JSON.stringify(path)}. Put the complete file text inside CDATA. No prose, Markdown fences, extra operands or additional instructions. To include the literal CDATA terminator ]]> in source, split it as ]]]]><![CDATA[>. Example envelope:\n${encodeSourceArtifact({ path, content: "complete file text" })}`;
}
